File: Apps/BankCore/backend/src/main/java/org/d3javu/backend/repository/base/CardRepository.java
package org.d3javu.backend.repository.base;

import org.d3javu.backend.model.base.card.Card;
import org.d3javu.backend.model.base.card.CardType;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.transaction.annotation.Isolation;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;


@Transactional(isolation = Isolation.SERIALIZABLE)
public interface CardRepository extends JpaRepository<Card, Long> {

    @Query(value = "insert into card (name, type, number, expiration_date, account_id, security_code) " +
            "values (:name, :type, :number, :expirationDate, :accountId, :securityCode) " +
            "returning id",
            nativeQuery = true)
    Long initCard(String name, String type, String number, LocalDate expirationDate, Long accountId, String securityCode);

    @Modifying
    @Query(value = "update card set number= :number where id= :cardId",
            nativeQuery = true)
    void updateCardAfterInit(Long cardId, String number);

    @Modifying
    @Query(value = "update card set name= :newName where id= :cardId",
            nativeQuery = true)
    void renameCard(Long cardId, String newName);

}

File: Apps/BankCore/backend/src/main/scala/repository/base/card/CardRepository.scala
package org.d3javu
package repository.base.card

import org.d3javu.domain.base.CardType
import org.d3javu.domain.base.PersonalAccount.AccountId

import domain.base.Card._

trait CardRepository[F[_]] {
  def initCard(
      name: CardName,
      cardType: CardType,
      number: CardNumber,
      expirationDate: ExpirationDate,
      accountId: AccountId,
      securityCode: SecurityCode,
  ): F[CardId]

  def updateCardAfterInit(cardId: CardId, number: CardNumber): F[Unit]

  def renameCard(cardId: CardId, newName: CardName): F[Unit]
}
